src/comgr.cpp

```diff
@@ -0,0 +1,503 @@
+                           const std::string& content,
+                           const amd_comgr_data_kind_t type)
+{
+    const Data d(type);
+    MIOPEN_LOG_I2(name << ' ' << content.size() << " bytes");
+    d.SetName(name);
+    d.SetBytes(content);
+    dataset.AddData(d);
+#if DEBUG_DETAILED_LOG
+    if(miopen::IsLogging(miopen::LoggingLevel::Info) && type == AMD_COMGR_DATA_KIND_SOURCE)
+    {
+        constexpr auto SHOW_FIRST = 1024;
+        const auto text_length    = (content.size() > SHOW_FIRST) ? SHOW_FIRST : content.size();
+        const std::string text(content, 0, text_length);
+        MIOPEN_LOG_I(text);
+    }
+#endif
+}
+
+void BuildOcl(const std::string& name,
+              const std::string& text,
+              const std::string& options,
+              const std::string& device,
+              std::vector<char>& binary)
+{
+    static const auto once = PrintVersion(); // Nice to see in the user's logs.
+    std::ignore            = once;
+
+    try
+    {
+        const Dataset inputs;
+        DatasetAddData(inputs, name, text, AMD_COMGR_DATA_KIND_SOURCE);
+        const ActionInfo action;
+        action.SetLanguage(AMD_COMGR_LANGUAGE_OPENCL_2_0);
+        const auto isaName = compiler::lc::GetIsaName(device);
+        MIOPEN_LOG_I2(isaName);
+        action.SetIsaName(isaName);
+        action.SetLogging(true);
+
+        auto optCompile = SplitSpaceSeparated(options);
+        compiler::lc::RemoveSuperfluousOptions(optCompile);
+        compiler::lc::AddOcl20CompilerOptions(optCompile);
+        action.SetOptionList(optCompile);
+
+        const Dataset addedPch;
+        action.Do(AMD_COMGR_ACTION_ADD_PRECOMPILED_HEADERS, inputs, addedPch);
+        const Dataset compiledBc;
+        action.Do(AMD_COMGR_ACTION_COMPILE_SOURCE_TO_BC, addedPch, compiledBc);
+
+        OptionList optLink;
+        optLink.push_back("wavefrontsize64");
+        for(const auto& opt : optCompile)
+        {
+            if(opt == "-cl-fp32-correctly-rounded-divide-sqrt")
+                optLink.push_back("correctly_rounded_sqrt");
+            else if(opt == "-cl-denorms-are-zero")
+                optLink.push_back("daz_opt");
+            else if(opt == "-cl-finite-math-only" || opt == "cl-fast-relaxed-math")
+                optLink.push_back("finite_only");
+            else if(opt == "-cl-unsafe-math-optimizations" || opt == "-cl-fast-relaxed-math")
+                optLink.push_back("unsafe_math");
+            else
+            {
+            } // nop
+        }
+        action.SetOptionList(optLink);
+        const Dataset addedDevLibs;
+        action.Do(AMD_COMGR_ACTION_ADD_DEVICE_LIBRARIES, compiledBc, addedDevLibs);
+        const Dataset linkedBc;
+        action.Do(AMD_COMGR_ACTION_LINK_BC_TO_BC, addedDevLibs, linkedBc);
+
+        action.SetOptionList(optCompile);
+        const Dataset relocatable;
+        action.Do(AMD_COMGR_ACTION_CODEGEN_BC_TO_RELOCATABLE, linkedBc, relocatable);
```

Comment:
Are you not checking compilation phases correct execution intentionally?

Reply:
`ActionInfo::Do()` checks comgr error codes and throws a `ComgrError` exception.